sdks/python/apache_beam/io/parquetio_test.py

```diff
@@ -0,0 +1,376 @@
+            num_shards=1, shard_name_template='')
+      with TestPipeline() as p:
+        # json used for stable sortability
+        readback = \
+            p \
+            | ReadFromParquet(path + '*') \
+            | Map(json.dumps)
+        assert_that(readback, equal_to([json.dumps(r) for r in self.RECORDS]))
+
+  def test_read_reentrant(self):
+    file_name = self._write_data()
+    source = _create_parquet_source(file_name)
+    source_test_utils.assert_reentrant_reads_succeed((source, None, None))
+
+  def test_read_without_splitting_multiple_row_group(self):
+    file_name = self._write_data(count=12000)
+    expected_result = self.RECORDS * 2000
+    self._run_parquet_test(file_name, None, None, False, expected_result)
+
+  def test_read_with_splitting_multiple_row_group(self):
+    file_name = self._write_data(count=12000)
+    expected_result = self.RECORDS * 2000
+    self._run_parquet_test(file_name, None, 10000, True, expected_result)
+
+  def test_split_points(self):
+    file_name = self._write_data(count=12000, row_group_size=3000)
+    source = _create_parquet_source(file_name)
+
+    splits = [
+        split
+        for split in source.split(desired_bundle_size=float('inf'))
+    ]
+    assert len(splits) == 1
+
+    range_tracker = splits[0].source.get_range_tracker(
+        splits[0].start_position, splits[0].stop_position)
+
+    split_points_report = []
+
+    for _ in splits[0].source.read(range_tracker):
+      split_points_report.append(range_tracker.split_points())
+
+    # There are a total of four row groups. Each row group has 3000 records.
+
+    # When reading records of the first group, range_tracker.split_points()
+    # should return (0, iobase.RangeTracker.SPLIT_POINTS_UNKNOWN)
+    self.assertEquals(
+        split_points_report[:10],
+        [(0, RangeTracker.SPLIT_POINTS_UNKNOWN)] * 10)
+
+    # When reading records of last group, range_tracker.split_points() should
+    # return (3, 1)
+    self.assertEquals(split_points_report[-10:], [(3, 1)] * 10)
+
+  def test_selective_columns(self):
+    file_name = self._write_data()
+    expected_result = [{'name': r['name']} for r in self.RECORDS]
+    self._run_parquet_test(file_name, ['name'], None, False, expected_result)
+
+  def test_sink_transform_multiple_row_group(self):
+    with tempfile.NamedTemporaryFile() as dst:
+      path = dst.name
+      with TestPipeline() as p:
+        # pylint: disable=expression-not-assigned
+        p \
+        | Create(self.RECORDS * 2000) \
+        | WriteToParquet(
+            path, self.SCHEMA, num_shards=1, codec='snappy',
+            shard_name_template='', row_group_size=3000)
+      self.assertEqual(pq.read_metadata(path).num_row_groups, 4)
+
+  def test_read_all_from_parquet_single_file(self):
+    path = self._write_data()
+    with TestPipeline() as p:
+      assert_that(
+          p \
+          | Create([path]) \
+          | ReadAllFromParquet(),
+          equal_to(self.RECORDS))
+
```

Comment:
Please consider adding exhausting tests for dynamic work rebalancing by using source_test_utils.assert_split_at_fraction_exhaustive similar to https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/textio_test.py#L310.